src/supported-operations/running-monitoring-operations.md

```diff
@@ -0,0 +1,94 @@
+* The device owner should be able to define constraints
+  when a command can be launched / stopped:
+  - Can two instances of the same operation run concurrently?
+  - Can the command run concurrently with another?
+  - Has the process to be detached from the agent daemon?
+  - Has an operation to be aborted on agent restart?
+
+## Proposed design
+* The operations are defined by the configuration files in [`/etc/tedge/operations`](./principles.md).
+* The cloud mappers trigger the operations.
+  * This is done on reception of operation requests from the cloud:
+    each operation configuration defines the topic to listen, and the message-pattern to match.
+  * The responsibility of the mapper is only to translate cloud specific messages into thin-edge requests.
+    When an operation request is received, the mapper combines the request and the operation configuration
+    into execution information, and delegates the execution to the agent.
+* The agent is the process that runs and monitors the operation commands.
+  * The agent listen for operation request on the `tedge/operations/start` topic.
+  * Each request provides the information to execute and monitor the operation:
+     * a request unique identifier,
+     * the operation name,
+     * the command path,
+     * an array of command arguments (at least the message that triggered the request),  
+     * the user running the command.
+  * ```json
+    { "id": "abcdef",
+      "operation": "c8y_LogfileRequest",
+      "command": "/usr/bin/c8y_logfile",
+      "args": ["522,DeviceSerial,logfileA,2013-06-22T17:03:14.000+02:00,2013-06-22T18:03:14.000+02:00,ERROR,1000"],
+      "user": "tedge"
+    }
+    ```
+* The agent schedules the operations reporting the progress on the `tedge/operations/status` topic.
+  * ```json
+    { "id": "abcdef",
+      "operation": "c8y_LogfileRequest",
+      "status": "scheduled"
+    }
+    ```
+  * `"status": "schedule"`: on reception of a well-formatted operation request.
+  * `"status": "executing"`: just before executing the operation.
+  * `"status": "successful"`: when the operation returns with a zero exit status.
+  * `"status": "failed"`: when the operation returns with a non-zero status.
+  * NOTE: One might also consider `cancelled` and `aborted` operations to deal with user cancellation
+    and operations killed by a signal or a power failure.
+* The mapper might subscribe to the `tedge/operations/status` topic
+  and translate the operation status into cloud specific messages.
+  * NOTE: There is open question here.
+    * Does the c8y mapper translate the `tedge/operations/status` messages into equivalent smartRest2 messages?
+    * Or does the c8y operation publish their progress directly to `c8y/s/us`?
+    * The former would be simpler for the operator implementor. 
+      However, most operations will in any case have to dialog over `c8y/s/us` and `c8y/s/ds`
+      to implement the specificities of the operation protocol.
```

Comment:
I know this may not be a welcome comment but I believe the best option would to leave to operation configuration and introduce a field there (eg: `report_status = true/false`) allowing for the user to specify which variant they want.

This would allow for more complex operations (maybe some daemon) which actually do extended communication to do that on their own but also allow for very simple operations to be implemented in "few line of code" (I'm thinking 
 operations which may do some house keeping or configuration management where the operation plugin doesn't need to do any communication and could be kept free of comms dependencies) and not to worry about communication

Reply:
And why not [rustbreak](https://github.com/TheNeikos/rustbreak) since Marcel in onboard now!

Reply:
Adding a parameter to the operation config is a valid option, even if I have a preference for a mechanism that is general and can be made specific combining small components.

Here, I would prefer:

* The mapper invoking a command for an operation and delegating all the c8y protocol to that process.
* The same mapper invoking a command that happens to be it-self a wrapper that handles the c8y process but delegates the actual work to the actual bare operation implementation - the one implemented in "few lines of code".